scripts: release: list_backports.py

cfriedt opened the pull request:
* create `list_backports.py` to examine prs applied to a backport branch and extract associated issues
  * helpful for adding to release notes
* add github action to verify backported changes have an associated issue (CI workflow added)
* use `GITHUB_TOKEN` in `bug_bash.py` and `list_issues.py`
* change `bug_bash.py` to use more consistent variable names (e.g. `--end END_DATE`)

```
Usage:
    # generate release notes for v2.7, with a date range, excluding specific PRs, output ReST
    ./scripts/release/list_backports.py \
        -t ~/.ghtoken \
        -b v2.7-branch \
        -s 2021-12-15 -e 2022-04-22 \
        -P 45074 -P 45868 -P 44918 -P 41234 -P 41174

    # verify a specific PR to v3.0 has an associated issue, output JSON
    GITHUB_TOKEN="<secret>" \
    ./scripts/release/list_backports.py \
        -b v3.0-branch \
        -p 43381 \
        -j | jq . | tee /tmp/backports.json
```

stephanosio commented on scripts/release/bug_bash.py:
```diff
@@ -26,39 +26,39 @@ def parse_args():
     parser.add_argument('-t', '--token', dest='tokenfile',
-                        help='File containing GitHub token', metavar='FILE')
-    parser.add_argument('-b', '--begin', dest='begin', help='begin date (YYYY-mm-dd)',
+                        help='File containing GitHub token (altnernatively, use GH_TOKEN env variable)', metavar='FILE')
```
typo

```suggestion
                        help='File containing GitHub token (alternatively, use GH_TOKEN env variable)', metavar='FILE')
```

stephanosio commented on scripts/release/bug_bash.py:
```diff
@@ -26,39 +26,39 @@ def parse_args():
     else:
-        if 'GITHUB_TOKEN' not in os.environ:
+        if 'GH_TOKEN' not in os.environ:
             raise ValueError('No credentials specified')
-        token = os.environ['GITHUB_TOKEN']
+        token = os.environ['GH_TOKEN']
```
Any reason behind this change? FWIW, GitHub calls it `GITHUB_TOKEN` (`${{ secrets.GITHUB_TOKEN }}`) and there are other scripts that use this name as well (see https://github.com/zephyrproject-rtos/zephyr/pull/45540).

cfriedt replied:
Ah weird - I thought I saw another script that used this convention. Have not been doing ci much lately

stephanosio replied:
https://github.com/zephyrproject-rtos/zephyr/blob/21d1ad3762302b3e461953df59430c77e0709274/scripts/release/list_issues.py#L150

right, there is `list_issues.py` that uses `GH_TOKEN`. Could you instead update this file to use `GITHUB_TOKEN` here?

cfriedt commented:
* accumulate PRs to ignore with `-x / --exclude-pull` option
* order pulls numerically in report

cfriedt commented:
* fix typo
* correct `GH_TOKEN` to `GITHUB_TOKEN` in `list_issues.py`

stephanosio commented on scripts/release/bug_bash.py:
```diff
@@ -26,17 +26,17 @@ def parse_args():
-                        help='File containing GitHub token', metavar='FILE')
-    parser.add_argument('-b', '--begin', dest='begin', help='begin date (YYYY-mm-dd)',
-                        metavar='date', type=valid_date_type, required=True)
+                        help='File containing GitHub token (altnernatively, use GITHUB_TOKEN env variable)', metavar='FILE')
```
```suggestion
                        help='File containing GitHub token (alternatively, use GITHUB_TOKEN env variable)', metavar='FILE')
```

mbolivar-nordic commented on scripts/release/bug_bash.py:
```diff
@@ -26,17 +26,17 @@ def parse_args():
-                        help='File containing GitHub token', metavar='FILE')
-    parser.add_argument('-b', '--begin', dest='begin', help='begin date (YYYY-mm-dd)',
-                        metavar='date', type=valid_date_type, required=True)
+                        help='File containing GitHub token (alternatively, use GITHUB_TOKEN env variable)', metavar='FILE')
```
These changes really feel like they should be part of a separate commit.

mbolivar-nordic commented on scripts/release/list_issues.py:
```diff
@@ -147,10 +147,10 @@ def parse_args():
     parse_args()
 
-    token = os.environ.get('GH_TOKEN', None)
+    token = os.environ.get('GITHUB_TOKEN', None)
```
These changes are not mentioned in the commit log and also seem like they should be in their own commit.

cfriedt replied:
I was asked to include them here by Stephanos. Would be fine to mention it in the commit msg.

mbolivar-nordic replied:
I see that the change was requested, but I see no request (or reason for one) that it has to be in this same commit; is there one?

fabiobaltieri commented on scripts/release/list_backports.py:
```diff
@@ -0,0 +1,218 @@
+        s += 'The following PRs are lacking associated issues:'
+        for p in pulls_without_issues:
+            s += f'\nhttps://github.com/zephyrproject-rtos/zephyr/pull/{p.number}'
+        raise ValueError(s)
```
Would it make sense to use a print and explicit exit? It's a bit odd to see a backtrace when this kicks in but it may be just me... up to you really.

Commit:
scripts: release: use GITHUB_TOKEN and start_date in scripts

Updated bug_bash.py and list_issues.py to use the GITHUB_TOKEN
environment variable for consistency with other scripts.

Updated bug_bash.py to use `-s / --start-date` instead of
`-b / --begin-date`.

Signed-off-by: Christopher Friedt <cfriedt@fb.com>

Commit:
scripts: release: list_backports.py

Created list_backports.py to examine prs applied to a backport
branch and extract associated issues. This is helpful for
adding to release notes.

The script may also be used to ensure that backported changes
also have one or more associated issues.

Signed-off-by: Christopher Friedt <cfriedt@fb.com>

Commit:
ci: backports: check if a backport PR has a valid issue

This is an automated check for the Backports project to
require one or more `Fixes #<issue>` items in the body
of the pull request.

Fixes #46164

Signed-off-by: Christopher Friedt <cfriedt@fb.com>

cfriedt commented on scripts/release/list_backports.py:
```diff
@@ -0,0 +1,218 @@
+def main():
+    args = parse_args()
+    bp = Backport(Github(args.token), args.base,
+                  args.start, args.end, set(args.excludes))
```
Fixed!

cfriedt commented:
@stephanosio - PTAL at `.github/workflows/backport-has-issue.yml`

stephanosio commented on .github/workflows/backport-has-issue.yml:
```diff
@@ -0,0 +1,21 @@
+name: Backport has Issue
+on:
+  pull_request_target:
+    on:
+      push:
+        branches:
```
This does not seem right. Have you tested this?

AFAIK, `branches` should be directly under `pull_request_target`.

cfriedt replied:
Fixed!

cfriedt replied:
Fixed!

stephanosio commented on scripts/release/list_backports.py:
```diff
@@ -0,0 +1,329 @@
+    def __init__(self, gh, base, pulls):
+        self._gh = gh
+        self._base = base
+        self._repo = gh.get_repo('zephyrproject-rtos/zephyr')
```
the org and repo names need to be configurable via the script parameters so that this script and workflow can be reused in the forks.

also they should be explicitly specified in the workflow based on the execution context:
https://github.com/zephyrproject-rtos/zephyr/blob/51c7777529a1a28b1e5ab5569cf6bf0cea7bd2df/.github/workflows/assigner.yml#L35-L36

cfriedt replied:
Fixed!

stephanosio commented on .github/workflows/backport-has-issue.yml:
```diff
@@ -0,0 +1,21 @@
+name: Backport has Issue
```
Please name this "Backport Issue Check" or something along the line.

cfriedt replied:
Fixed!

stephanosio commented:
Marking this DNM until the CI workflow is fully tested (will get to this some time next week).

fabiobaltieri commented on scripts/release/list_backports.py:
```diff
@@ -0,0 +1,340 @@
+
+        unfiltered_pulls = repo.get_pulls(
+            base=base, state='closed')
+        for p in unfiltered_pulls:
```
These takes forever for me, are the APIs just that slow? Wondering if it's worth adding a verbose flag to print `unfiltered_pulls.totalCount` here and then a line as PR get fetched, what do you think? (can't see any way of making it faster but maybe I'm missing something)

fabiobaltieri replied:
Yeah it does not look like it's possible using that API right now, though `list_issues.py` does that using `requests.get` directly for issues: https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/release/list_issues.py#L109, worth looking into it? Probably not worth it for something we run once per release.

cfriedt replied:
It might be worth optimizing that at some point later, or maybe requesting GH add that feature to their API, but yeah, probably not pressing right now.

cfriedt commented:
> Marking this DNM until the CI workflow is fully tested (will get to this some time next week).

@stephanosio - any update?

stephanosio commented on scripts/release/list_backports.py:
```diff
@@ -0,0 +1,340 @@
+            obj = {}
+            obj['id'] = i.number
+            obj['title'] = Backport.sanitize_title(i.title)
+            obj['url'] = f'https://github.com/zephyrproject-rtos/zephyr/pull/{i.number}'
```
need to use the org and repo names from the arg.

cfriedt replied:
Fixed!

stephanosio commented on scripts/release/list_backports.py:
```diff
@@ -0,0 +1,340 @@
+                    match = re.search(r"^Fixes[:]?\s*#([1-9][0-9]*).*", line)
+                    if not match:
+                        match = re.search(
+                            r"^Fixes[:]?\s*https://github\.com/zephyrproject-rtos/zephyr/issues/([1-9][0-9]*).*", line)
```
same here

cfriedt replied:
Fixed!

stephanosio commented on scripts/release/list_backports.py:
```diff
@@ -0,0 +1,340 @@
+                            self._pulls_with_invalid_issues[p.number].append(
+                                issue_number)
+                        logging.error(
+                            f'https://github.com/zephyrproject-rtos/zephyr/pull/{p.number} references invalid issue number {issue_number}')
```
same here

cfriedt replied:
Fixed!

stephanosio commented on scripts/release/list_backports.py:
```diff
@@ -0,0 +1,340 @@
+            # report prs missing issues later
+            if len(issues_for_this_pr) == 0:
+                logging.error(
+                    f'https://github.com/zephyrproject-rtos/zephyr/pull/{p.number} does not have an associated issue')
```
ditto

cfriedt replied:
Fixed!

stephanosio commented on scripts/release/list_backports.py:
```diff
@@ -0,0 +1,340 @@
+        logging.error('The following PRs link to invalid issues:')
+        for (p, lst) in pulls_with_invalid_issues:
+            logging.error(
+                f'\nhttps://github.com/zephyrproject-rtos/zephyr/pull/{p.number}: {lst}')
```
ditto

cfriedt replied:
Fixed!

stephanosio commented on scripts/release/list_backports.py:
```diff
@@ -0,0 +1,340 @@
+        logging.error('The following PRs are lacking associated issues:')
+        for p in pulls_without_issues:
+            logging.error(
+                f'https://github.com/zephyrproject-rtos/zephyr/pull/{p.number}')
```
ditto

cfriedt replied:
Fixed!

stephanosio reviewed (changes requested):
Please update the workflow file as follows:
https://github.com/zephyrproject-rtos/zephyr-testing/blob/v9.9-branch/.github/workflows/backport_issue_check.yml

stephanosio commented:
@cfriedt also seeing some errors running the script:
```
$ scripts/release/list_backports.py -o zephyrproject-rtos -r zephyr-testing -b v9.9-branch -p 176
Traceback (most recent call last):
  File "scripts/release/list_backports.py", line 340, in <module>
    sys.exit(main())
  File "scripts/release/list_backports.py", line 313, in main
    pulls_with_invalid_issues = bp.get_pulls_with_invalid_issues()
  File "scripts/release/list_backports.py", line 280, in get_pulls_with_invalid_issues
    self.get_issues()
  File "scripts/release/list_backports.py", line 257, in get_issues
    issue_map = issue_map | issues_for_this_pr
TypeError: unsupported operand type(s) for |: 'dict' and 'dict'
```

https://github.com/zephyrproject-rtos/zephyr-testing/pull/176

cfriedt commented:
> @cfriedt also [seeing some errors running the script](https://github.com/zephyrproject-rtos/zephyr-testing/runs/7362938582?check_suite_focus=true#step:4:11):
> ```
> TypeError: unsupported operand type(s) for |: 'dict' and 'dict'
> ```

@stephanosio - oh wow - is it possible that my python skills are more bleeding edge than Zephyr can handle? (Kidding)

That operator is for appending one dict to another, but I guess only for python 3.9+

https://stackoverflow.com/a/38176082

cfriedt commented:
> @cfriedt also [seeing some errors running the script](https://github.com/zephyrproject-rtos/zephyr-testing/runs/7362938582?check_suite_focus=true#step:4:11):
> 
> ```
> $ scripts/release/list_backports.py -o zephyrproject-rtos -r zephyr-testing -b v9.9-branch -p 176
> Traceback (most recent call last):
>   File "scripts/release/list_backports.py", line 340, in <module>
>     sys.exit(main())
>   File "scripts/release/list_backports.py", line 313, in main
>     pulls_with_invalid_issues = bp.get_pulls_with_invalid_issues()
>   File "scripts/release/list_backports.py", line 280, in get_pulls_with_invalid_issues
>     self.get_issues()
>   File "scripts/release/list_backports.py", line 257, in get_issues
>     issue_map = issue_map | issues_for_this_pr
> TypeError: unsupported operand type(s) for |: 'dict' and 'dict'
> ```
> 
> [zephyrproject-rtos/zephyr-testing#176](https://github.com/zephyrproject-rtos/zephyr-testing/pull/176)

cfriedt commented on .github/workflows/backport_issue_check.yml:
```diff
@@ -0,0 +1,21 @@
+name: Backport Issue Check
+on:
+  pull_request_target:
+    branches:
+      - v(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)-branch
+      - v(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)-auditable-branch
```
Fixed!

stephanosio commented on .github/workflows/backport_issue_check.yml:
```diff
@@ -0,0 +1,20 @@
+name: Backport Issue Check
```
Please update the workflow as well: https://github.com/zephyrproject-rtos/zephyr/pull/45880#pullrequestreview-1040626430

Reworked workflow: https://github.com/zephyrproject-rtos/zephyr-testing/blob/v9.9-branch/.github/workflows/backport_issue_check.yml

You can copy and paste the content of the linked workflow file above, which has been verified in the testing repo, into this PR.

cfriedt replied:
Done!

stephanosio reviewed (approved):
Tested in https://github.com/zephyrproject-rtos/zephyr-testing/pull/176

Without linked issue: https://github.com/zephyrproject-rtos/zephyr-testing/runs/7368173127?check_suite_focus=true#step:4:12
With linked issue: https://github.com/zephyrproject-rtos/zephyr-testing/runs/7368167432?check_suite_focus=true#step:4:11

cfriedt commented:
It's weird that CI still seems to think this PR has a DNM label
![Screenshot_20220716-062245.png](https://user-images.githubusercontent.com/983494/179351027-baa8fb4a-7211-42a1-8bfb-4b5b16e04a8d.png)

cfriedt commented:
> It's weird that CI still seems to think this PR has a DNM label
> ![Screenshot_20220716-062245.png](https://user-images.githubusercontent.com/983494/179351027-baa8fb4a-7211-42a1-8bfb-4b5b16e04a8d.png)

Yeah, @stephanosio - it seems removing the DNM label and re-running the job still doesn't pass CI - it seems like kind of a bug, no?

stephanosio commented:
> Yeah, @stephanosio - it seems removing the DNM label and re-running the job still doesn't pass CI - it seems like kind of a bug, no?

You might need to refresh. For me, "Do Not Merge / Prevent Merging (pull_request)" is shown as skipped.

stephanosio commented:
NOTE: This needs to be "backported" to the relevant release branches to work properly.

v2.7: https://github.com/zephyrproject-rtos/zephyr/pull/47951
v3.0: https://github.com/zephyrproject-rtos/zephyr/pull/47953
v3.1: https://github.com/zephyrproject-rtos/zephyr/pull/47952

cfriedt commented:
> NOTE: This needs to be "backported" to the relevant release branches to work properly.
> 
> v2.7: #47951 v3.0: #47953 v3.1: #47952

Thanks for merging that into version branches @stephanosio ;-)
